Make shuffling optional in DistributedSampler

dehuacheng opened the pull request:
Summary:
In some cases, for example, when we training on CTR data, we would like to start training from old samples and finish on new recent samples.

This diff add the option to disable the shuffling in DistributedSampler to accommodate this use case.

Differential Revision: D16100388

soumith commented on torch/utils/data/distributed.py:
```diff
@@ -37,12 +38,17 @@ def __init__(self, dataset, num_replicas=None, rank=None):
         g.manual_seed(self.epoch)
-        indices = torch.randperm(len(self.dataset), generator=g).tolist()
+        indices = (
+            torch.randperm(len(self.dataset), generator=g).tolist()
```
break this out into a separate if/else block, instead inlining if/else inside indices = (

soumith commented:
I'm also really confused. If you dont need random shuffling, you dont need a DistributedSampler at all, right?
It will just be a pure iterator with no shuffling

Commit:
Make shuffling optional in DistributedSampler (#22479)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22479

In some cases, for example, when we training on CTR data, we would like to start training from old samples and finish on new recent samples.

This diff add the option to disable the shuffling in DistributedSampler to accommodate this use case.

Differential Revision: D16100388

fbshipit-source-id: e0d31328920a4aa2315820683242cc44959c5d82

soumith commented:
okay, nevermind, this actually makes sense to effectively do sharding without shuffling. It took me a second to wrap my head around this

SsnL commented:
test?

soumith commented:
my bad on the code review. i should've requested a test
